_universum/modules/vcs/perforce_vcs.py

```diff
@@ -560,6 +560,15 @@ def clean_workspace(self):
             self.p4.client = self.client_name
             report = self.p4.run_revert("//...")
             self.p4report(report)
+            shelves = p4.run_changes("-c", self.client_name, "-s", "shelved")
+            for item in shelves:
+                self.out.log("Deleting shelve from CL " + item["change"])
+                p4.run_shelve("-d" , "-c", item["change"])
+            self.p4.run_revert("//...")
```

Comment:
I this revert really needed? As far as I understand you have already reverted repo in the line 561.

Reply:
Yes. 'p4 shelve -d -c 1234' will delete any changes, stored in CL 1234 _after_ unshelving these changes into CL itself. That's why _before_ this command CL contents must be already reverted to prevent resolving; and that's why this CL should be reverted again after.
This can be avoided by using '-f' option, but it is not available to non-admin users.

Reply:
Are you sure? 
I can remove the shelve without any "auto-unshelving" in the GUI version of P4. Command is the same "p4 shelve -d -c 18381526"

Application: 
Helix Visual Client version: P4V/NTX64/2019.2/1904275

Reply:
For more information please refer to official guide:
https://www.perforce.com/manuals/v18.1/cmdref/index.html#CmdRef/p4_shelve.html
('-d' option)

Reply:
Ok, thank you

Reply:
I just checked this behavior on real perforce instance.
There is no issue with having the same file in pending state (being checked out) while deleting its shelved version.
`p4 shelve -d -c 1234` correctly deleted shelve and didn't change the checked out file.

Reply:
The decision to keep second revert has to be documented somewhere. This is because the p4 documentation is counterintuitive and doesn't correspond to implementation.